### PR TITLE
feat: support for lookarounds and non-capture groups

### PR DESCRIPTION
**Pull request by PaulJPhilp**
### Summary

Adding support for look-arounds:
   - positiveLookahead
   - negativeLookahead
   - positiveLookbehind
   - negativeLookbehind

Adding support for non-capture groups.

**Checklist**
- [x] Implementation
- [x] Tests (pattern and matching)
- [x] API docs 
- [x] README docs
- [x] Example docs & tests

### Test plan

Adding one test file for each function:
   - positive-lookahead.test.ts, 
   - positive-lookbehind.test.ts, 
   - negative-lookbehind.test.ts, 
   - negative-lookahead.test.ts, 
   - non-capture-group.test.ts

### CODE REVIEW

File: docs/API.md [hunk 1]
```diff
@@ -58,6 +58,65 @@ Regex syntax: `(...)`.
 
 Captures, also known as capturing groups, extract and store parts of the matched string for later use.
 
+### `nonCaptureGroup()`
```
Comment (mdjastrzebski) [inline docs/API.md:61]: @PaulJPhilp Do you think we need explicit `nonCaptureGroups` as a feature. We already do it implicitly when needed (e.g. applying quantifiers like `zeroOrMore`) to non-atomic values. What would be the benefits of having this as an explicit feature?

File: docs/API.md [hunk 2]
```diff
@@ -58,6 +58,65 @@ Regex syntax: `(...)`.
...
+```
+
+Regex syntax: `(?:...)`.
+
+Groups multiple tokens together without creating a capture group.
+
+### `positiveLookahead()`
+
+```ts
+function positiveLookahead(
```
Comment (mdjastrzebski) [inline docs/API.md:76]: How about `lookahead` for positive one and `negetiveLookahead` for negative one, mimicking [Swift Regex Builder](https://github.com/apple/swift-evolution/blob/main/proposals/0351-regex-builder.md#anchors-and-lookaheads) in naming?

**Comment by mdjastrzebski** [comment]
@PaulJPhilp awesome that you contributed this PR. I've left some comments.

Could you also prepare some meaningful, real-life example of usage for lookaheads/look behinds? Like we have in `Examples.md` and `example-xxx.test.ts`.

**Comment by mdjastrzebski** [comment]
@PaulJPhilp before making a call on this I want to clarify the reasoning behind adding `nonCaptureGroup` is only to have it as explicit construct because people referencing regex expressions might be looking for it? Is there any other benefit for having it?

**Comment by PaulJPhilp** [comment]
I use it for structuring and debugging.  A form of a no-op.  I prefer it to
using '[Element1, Element2]' form but this is just a personal preference.
The main documented use case (which I haven't used) is for asserting a
repeated pattern which, as you say, can be accomplished in other ways.

I can support either decision but prefer completeness.  If it's not kept, I
think a note on alternatives would be helpful.

Paul

**Review by mdjastrzebski — APPROVED** [review]
Looks great! 🔥

I've decided to remove `nonCaptureGroup` but to specifically note in the docs that this is not needed, on order to resolve potential issues from people looking for it.

Huge thanks especially for really advanced examples.

**Pull request merged**